### ISSUE TITLE
Provide a shared library

### ISSUE THREAD

**Issue by GoogleCodeExporter** [chrsmithdemos/leveldb#27]
```
Please add a target into the Makefile to compile a shared library object.

Thanks in advance for any reply.

```

---

Original issue reported on code.google.com by `quadris...@gmail.com` on 9 Aug 2011 at 12:57

**Comment by GoogleCodeExporter** [comment]
```
Thanks for your suggestion quadris! Do you have a patch to make this happen or 
would you be interested in making one?
```

---

Original comment by `ga...@google.com` on 10 Aug 2011 at 9:11

**Comment by GoogleCodeExporter** [comment]
```
Thanks for the quick reply!

OK, I'll prepare a patch ASAP, it should be ready within few days.

As an aside, I've uploaded leveldb to Debian unstable and now waiting to see it 
joining sid, check this URL for more information about the status:
http://bugs.debian.org/cgi-bin/bugreport.cgi?bug=636695
```

---

Original comment by `quadris...@gmail.com` on 10 Aug 2011 at 10:41

**Comment by GoogleCodeExporter** [comment]
```
A patch is attached. Please note it might need some testing.
```

---

Original comment by `quadris...@gmail.com` on 11 Aug 2011 at 7:49

Attachments:
- [1001-buildsystem.patch](https://storage.googleapis.com/google-code-attachments/leveldb/issue-27/comment-3/1001-buildsystem.patch)

**Comment by GoogleCodeExporter** [comment]
```
It would be great if this patch was to be committed. For what it's worth, it's 
being shipped with OpenBSD's leveldb port.
```

---

Original comment by `jasperi...@gmail.com` on 1 Nov 2011 at 8:32

**Comment by GoogleCodeExporter** [comment]
```
Hi there - apologies for having let this issue linger so long. 

I've reached out to quadris for a new patch based on a more recent LevelDB 
version, and for the CLA so we can patch this in.

Gabor
```

---

Original comment by `ga...@google.com` on 1 Nov 2011 at 6:50
- Changed state: **Accepted**

**Comment by GoogleCodeExporter** [comment]
```
Hi,

sorry for the delay.

You find the patch attached, it's written against the most recent snapshot: 
36a5f8e
```

---

Original comment by `quadris...@gmail.com` on 8 Nov 2011 at 2:04

Attachments:
- [0010-shared_object.patch](https://storage.googleapis.com/google-code-attachments/leveldb/issue-27/comment-6/0010-shared_object.patch)

**Comment by GoogleCodeExporter** [comment]
```
Hi quadris, your patch to not override CFLAGS, etc. has been landed.

The code to build a shared library isn't portable (wouldn't work on Mac OS X 
for example), so I did not commit it. We need to figure out a portable way to 
do this.
```

---

Original comment by `h...@chromium.org` on 14 Nov 2011 at 5:25

**Comment by GoogleCodeExporter** [comment]
```
Hi, I've rigged up the shared library patch to work on OS X and Linux and 
attached it here. I don't have a Sun or FreeBSD machine to test it on.

There a CLA I need to sign or some such? (I freely give over my rights to this 
code, in case there isn't a CLA.)
```

---

Original comment by `j...@somethingsimilar.com` on 15 Jan 2012 at 8:39

Attachments:
- [shared_library.patch](https://storage.googleapis.com/google-code-attachments/leveldb/issue-27/comment-8/shared_library.patch)

**Comment by GoogleCodeExporter** [comment]
```
Also, I'm willing to rework this patch. Two things that suck: 
SHARED_INSTALL_PATH could probably be INSTALL_PATH and I was not able to figure 
out how to pass along the SHARED_MAJOR and SHARED_LIBRARY variables to 
build_detect_platform script, so there is some duplication.
```

---

Original comment by `j...@somethingsimilar.com` on 15 Jan 2012 at 8:42

**Comment by GoogleCodeExporter** [comment]
```
Here's an updated version of the patch. Two changes occured.

1) -f has been added to the ln commands that create the shared library 
symlinks. Without this, users were prevented from rebuilding the shard library 
with all without first running clean.

2) SHARED_INSTALL_PATH has been renamed to INSTALL_PATH. It's still not ideal, 
but I can't recall what the typical convention is for this.
```

---

Original comment by `j...@somethingsimilar.com` on 16 Jan 2012 at 11:17

Attachments:
- [shared_library.patch](https://storage.googleapis.com/google-code-attachments/leveldb/issue-27/comment-10/shared_library.patch)

**Comment by GoogleCodeExporter** [comment]
```
Here's another version with the recent changes to build_detect_platform merged 
in, and an install target added.
```

---

Original comment by `j...@somethingsimilar.com` on 11 Feb 2012 at 5:04

Attachments:
- [shared_library.patch](https://storage.googleapis.com/google-code-attachments/leveldb/issue-27/comment-11/shared_library.patch)

**Comment by GoogleCodeExporter** [comment]
```
Is there anything that prevents this from being merged upstream? Bindings 
authors (such as myself) would appreciate.
```

---

Original comment by `kim.alti...@gmail.com` on 17 Mar 2012 at 5:53

**Comment by GoogleCodeExporter** [comment]
```
Here's an updated version of the patch against the latest HEAD. It also adds 
passing LDFLAGS when building the shared lib, otherwise linking against snappy 
would fail.
```

---

Original comment by `kim.alti...@gmail.com` on 17 Mar 2012 at 8:32

Attachments:
- [shared_library.patch](https://storage.googleapis.com/google-code-attachments/leveldb/issue-27/comment-13/shared_library.patch)

**Comment by GoogleCodeExporter** [comment]
```
Updated again against master. This is a much cleaner version thanks to the 
recent changes in build_detect_platform. 

It also includes Kim's addition of LDFLAGS to the shared library target. I hope 
I didn't miss any other changes.
```

---

Original comment by `j...@somethingsimilar.com` on 25 Mar 2012 at 3:09

Attachments:
- [shared_library.patch](https://storage.googleapis.com/google-code-attachments/leveldb/issue-27/comment-14/shared_library.patch)

**Comment by GoogleCodeExporter** [comment]
```
The attached patch fixes the SONAME, which should not include any path.
```

---

Original comment by `quadris...@gmail.com` on 13 Apr 2012 at 9:44

Attachments:
- [0111-soname_fix.patch](https://storage.googleapis.com/google-code-attachments/leveldb/issue-27/comment-15/0111-soname_fix.patch)

**Comment by GoogleCodeExporter** [comment]
```
Please apply the as-needed patch attached to issue #57 too.
```

---

Original comment by `quadris...@gmail.com` on 13 Apr 2012 at 9:45

**Comment by GoogleCodeExporter** [comment]
```
Moreover, this adds an 'install' target to the Makefile.
Thanks for considering.
```

---

Original comment by `quadris...@gmail.com` on 13 Apr 2012 at 9:49

Attachments:
- [0112-makefile_install.patch](https://storage.googleapis.com/google-code-attachments/leveldb/issue-27/comment-17/0112-makefile_install.patch)

**Comment by GoogleCodeExporter** [comment]
```
A future release will fix the SONAME problem. Thanks.

We explicitly decided not to put installation support in Makefile since it
can vary widely across different platforms/users and we don't want the
Makefile to get cluttered with confusing/conditional/customizable rules for
installation.

```

---

Original comment by `san...@google.com` on 17 Apr 2012 at 9:22

**Comment by GoogleCodeExporter** [comment]
```
Explain the removal of the INSTALL_PATH for me, please? Without it, I'm unable 
to build and use the library in place on OS X.

dyld: Library not loaded: libleveldb.dylib.1
  Referenced from: /var/folders/r_/64kn6_z15s580w27scg2z1lw0000gn/T/go-build810548351/github.com/jmhodges/levigo/_test/levigo.test
  Reason: image not found

```

---

Original comment by `j...@somethingsimilar.com` on 22 Sep 2012 at 6:01

**Comment by GoogleCodeExporter** [comment]
```
We can leave it out, but OS X needs `PLATFORM_SHARED_LDFLAGS="-dynamiclib 
-Wl,-headerpad_max_install_names,-install_name,"` set in build_detect_platform 
and an install target to call  out to `install_name_tool -id  
/Users/jhodges/projects/leveldb/libleveldb.dylib.1 libleveldb.dylib.1`

That or we special case OS X to get the INSTALL_PATH and added it to 
PLATFORM_SHARED_LDFLAGS and include it in build_detect_platform. Currently, 
INSTALL_PATH is not visible to build_detect_platfrom.

```

---

Original comment by `j...@somethingsimilar.com` on 22 Sep 2012 at 6:23

**Comment by GoogleCodeExporter** [comment]
```
That should be `install_name_tool -id ${INSTALL_PATH}/libleveldb.dylib.1 
libleveldb.dylib.1`, of course.
```

---

Original comment by `j...@somethingsimilar.com` on 22 Sep 2012 at 6:48

**Comment by GoogleCodeExporter** [comment]
```
Fixed in 1.7.0.

```

---

Original comment by `san...@google.com` on 16 Oct 2012 at 11:32
- Changed state: **Fixed**
